# 1 - Preface

Kelompok 1

Nama : Ronan Lazuardi, Muhammad Fariz Firdaus, Syahrul Budi Rahmadan

Batch : FTDS-SBY-002

Objective : 

URL Dataset : [Kaggle](https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis/data)

# 2 - Import Libraries

In [99]:
import pandas as pd 
pd.set_option('display.max_columns', None)
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# 3 - Data Loading

In [76]:
# Loading file csv dataset raw

df_raw = pd.read_csv("marketing_campaign.csv", sep='\t')

Deskripsi tiap kolom/feature
|Feature|Description|
|:-|:-|
|AcceptedCmp1|1 if customer accepted the offer in the 1st campaign, 0 otherwise|
|AcceptedCmp2|1 if customer accepted the offer in the 2nd campaign, 0 otherwise|
|AcceptedCmp3|1 if customer accepted the offer in the 3rd campaign, 0 otherwise|
|AcceptedCmp4|1 if customer accepted the offer in the 4th campaign, 0 otherwise|
|Response (target)|1 if customer accepted the offer in the last campaign, 0 otherwise|
|Complain|1 if customer complained in the last 2 years|
|DtCustomer|Date of customer's enrollment with the company|
|Education|Customer's level of education|
|Marital|Customer's marital status|
|Kidhome|Number of small children in customer's household|
|Teenhome|Number of teenagers in customer's household|
|Income|Customer's yearly household income|
|MntFishProducts|Amount spent on fish products in the last 2 years|
|MntMeatProducts|Amount spent on meat products in the last 2 years|
|MntFruits|Amount spent on fruits in the last 2 years|
|MntSweetProducts|Amount spent on weet products in the last 2 years|
|MntWines|Amount spent on wines in the last 2 years|
|MntGoldProds|Amount spent on gold products in the last 2 years|
|NumDealsPurchases|Number of purchases made with discount|
|NumStorePurchases|Number of purchases made using catalogue|
|NumWebPurchases|Number of purchases made through company's web site|
|NumWebVisitsMonth|Number of visits to company's web site in the last month|
|Recency|Number of days since the last purchase|

In [77]:
# Copy untuk menjaga data asli
df = df_raw.copy()

In [78]:
# Menampilkan 5 baris awal
df.head(5)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
0  04-09-2012       58       635         88              546              172   
1  08-03-2014       38        11          1                6                2   
2  21-08-2013       26       426         49              127              111   
3  10-02-2014       26        11          4               20               10   
4  19-01-2014       94       173         43              118               46   

   MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0                88            88                  3                8   
1                 1             6                  2                1   
2                21            42                  1                8   
3                 3             5                  2                2   
4                27            15                  5                5   

   NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  \
0                   10                  4                  7             0   
1                    1                  2                  5             0   
2                    2                 10                  4             0   
3                    0                  4                  6             0   
4                    3                  6                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0

In [79]:
# Menampilkan 5 baris acak
df.sample(5)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
717   1715        1978  Graduation       Together  25851.0        1         0   
2236  4001        1946         PhD       Together  64014.0        2         1   
864   4120        1975    2n Cycle         Single  49514.0        1         0   
557   8314        1982  Graduation        Married  79908.0        0         0   
139   1461        1965         PhD       Divorced  36921.0        1         1   

     Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
717   11-02-2013       45         9          4               18   
2236  10-06-2014       56       406          0               30   
864   17-12-2013       61        88         39               78   
557   05-04-2013       30       557        129              761   
139   28-07-2013       74        17          1               12   

      MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
717                 7                 5             5                  2   
2236                0                 0             8                  7   
864                58                13            93                  2   
557                29               185           111                  1   
139                 3                 1             4                  2   

      NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
717                 2                    0                  3   
2236                8                    2                  5   
864                 6                    1                  4   
557                 5                    6                  7   
139                 2                    0                  3   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
717                   7             0             0             0   
2236                  7             0             0             0   
864                   7             0             0             0   
557                   2             0             0             1   
139                   7             0             0             0   

      AcceptedCmp1  AcceptedCmp2  Complain  Z_CostContact  Z_Revenue  Response  
717              0             0         0              3         11         0  
2236             1             0         0              3         11         0  
864              0             0         0              3         11         0  
557              0             0         0              3         11         0  
139              0             0         0              3         11         0

In [80]:
# Menampilkan 5 baris akhir
df.tail(5)

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
2235         1  13-06-2013       46       709         43              182   
2236         1  10-06-2014       56       406          0               30   
2237         0  25-01-2014       91       908         48              217   
2238         1  24-01-2014        8       428         30              214   
2239         1  15-10-2012       40        84          3               61   

      MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
2235               42               118           247                  2   
2236                0                 0             8                  7   
2237               32                12            24                  1   
2238               80                30            61                  2   
2239                2                 1            21                  3   

      NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
2235                9                    3                  4   
2236                8                    2                  5   
2237                2                    3                 13   
2238                6                    5                 10   
2239                3                    1                  4   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
2235                  5             0             0             0   
2236                  7             0             0             0   
2237                  6             0             1             0   
2238                  3             0             0             0   
2239                  7             0             0             0   

      AcceptedCmp1  AcceptedCmp2  Complain  Z_CostContact  Z_Revenue  Response  
2235             0             0         0              3         11         0  
2236             1             0         0              3         11         0  
2237             0             0         0              3         11         0  
2238             0             0         0              3         11         0  
2239             0             0         0              3         11         1

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [82]:
df.shape

(2240, 29)

Dataset 'marketing_campaign.csv' terdiri dari 2240 baris dan 29 kolom. Kolom dalam dataset memiliki beberapa tipe data yaitu:

1. Int64 : 25 kolom
2. Object : 3 kolom
3. Float64 : 1 kolom

In [83]:
for col in df.columns:
    print('Nilai unik', col, df[col].nunique())

Nilai unik ID 2240
Nilai unik Year_Birth 59
Nilai unik Education 5
Nilai unik Marital_Status 8
Nilai unik Income 1974
Nilai unik Kidhome 3
Nilai unik Teenhome 3
Nilai unik Dt_Customer 663
Nilai unik Recency 100
Nilai unik MntWines 776
Nilai unik MntFruits 158
Nilai unik MntMeatProducts 558
Nilai unik MntFishProducts 182
Nilai unik MntSweetProducts 177
Nilai unik MntGoldProds 213
Nilai unik NumDealsPurchases 15
Nilai unik NumWebPurchases 15
Nilai unik NumCatalogPurchases 14
Nilai unik NumStorePurchases 14
Nilai unik NumWebVisitsMonth 16
Nilai unik AcceptedCmp3 2
Nilai unik AcceptedCmp4 2
Nilai unik AcceptedCmp5 2
Nilai unik AcceptedCmp1 2
Nilai unik AcceptedCmp2 2
Nilai unik Complain 2
Nilai unik Z_CostContact 1
Nilai unik Z_Revenue 1
Nilai unik Response 2


In [84]:
# Drop kolom 'ID', 'Z_CostContact', 'Z_Revenue'

df.drop(columns=['ID','Z_CostContact','Z_Revenue'],inplace=True)

Kolom `ID`, `Z_CostContact`, `Z_Revenue` kami **drop** dikarenakan
kolom `ID` merupakan primary key / kolom identitas unik customer dan memiliki nilai unik di setiap baris,
sedangkan kolom `Z_CostContact` dan `Z_Revenue` memiliki 1 nilai yang sama sepanjang kolom,
sehingga ketiga kolom ini tidak relevan untuk dianalisa.

In [87]:
# Cek data duplikat

df.duplicated().sum()

182

Ada 182 baris di dalam dataset yang merupakan baris dengan nilai yang terduplikasi. Baris terduplikat akan dibuang karena akan mengganggu hasil prediksi nantinya.

In [93]:
# Menghapus baris duplikat

df.drop_duplicates(inplace=True)

In [94]:
# Cek data NaN / missing value

df.isna().sum()

Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Response                0
dtype: int64

Ada 24 baris di dalam dataset yang merupakan baris pada kolom `Income` yang tidak memiliki nilai yaitu NaN atau Null atau Missing Value. Baris dengan missing value ini akan dibuang karena akan mengganggu hasil prediksi nantinya.

In [95]:
# Menghapus baris NaN / Missing Value

df.dropna(inplace=True)

In [97]:
# Cek shape dataset setelah drop kolom tidak relevan, baris duplikat & missing value

df.shape

(2034, 26)

Setelah drop kolom yang tidak relevan dan baris-baris yang terduplikat dan missing value, kini dataset menjadi terdiri dari 2034 baris dan 26 kolom.

In [98]:
df.describe().T

count          mean           std     min       25%  \
Year_Birth           2034.0   1968.802852     11.975420  1893.0   1959.00   
Income               2034.0  52357.791544  25526.956988  1730.0  35528.25   
Kidhome              2034.0      0.442970      0.535914     0.0      0.00   
Teenhome             2034.0      0.508358      0.546130     0.0      0.00   
Recency              2034.0     48.844641     28.983678     0.0     24.00   
MntWines             2034.0    305.180924    337.323274     0.0     23.00   
MntFruits            2034.0     26.292527     39.772885     0.0      2.00   
MntMeatProducts      2034.0    167.763520    226.790697     0.0     16.00   
MntFishProducts      2034.0     37.449361     54.792015     0.0      3.00   
MntSweetProducts     2034.0     27.141101     41.494430     0.0      1.00   
MntGoldProds         2034.0     43.809735     51.741969     0.0      9.00   
NumDealsPurchases    2034.0      2.337758      1.933849     0.0      1.00   
NumWebPurchases      2034.0      4.099312      2.756911     0.0      2.00   
NumCatalogPurchases  2034.0      2.670600      2.937896     0.0      0.00   
NumStorePurchases    2034.0      5.782203      3.238853     0.0      3.00   
NumWebVisitsMonth    2034.0      5.322517      2.438665     0.0      3.00   
AcceptedCmp3         2034.0      0.074238      0.262222     0.0      0.00   
AcceptedCmp4         2034.0      0.076205      0.265390     0.0      0.00   
AcceptedCmp5         2034.0      0.072271      0.259000     0.0      0.00   
AcceptedCmp1         2034.0      0.065388      0.247271     0.0      0.00   
AcceptedCmp2         2034.0      0.012783      0.112363     0.0      0.00   
Complain             2034.0      0.009833      0.098696     0.0      0.00   
Response             2034.0      0.153392      0.360454     0.0      0.00   

                         50%       75%       max  
Year_Birth            1970.0   1977.00    1996.0  
Income               51533.0  68480.75  666666.0  
Kidhome                  0.0      1.00       2.0  
Teenhome                 0.0      1.00       2.0  
Recency                 49.0     74.00      99.0  
MntWines               175.5    505.00    1493.0  
MntFruits                8.0     33.00     199.0  
MntMeatProducts         68.0    230.00    1725.0  
MntFishProducts         12.0     50.00     259.0  
MntSweetProducts         8.0     33.75     262.0  
MntGoldProds            24.0     56.00     321.0  
NumDealsPurchases        2.0      3.00      15.0  
NumWebPurchases          4.0      6.00      27.0  
NumCatalogPurchases      2.0      4.00      28.0  
NumStorePurchases        5.0      8.00      13.0  
NumWebVisitsMonth        6.0      7.00      20.0  
AcceptedCmp3             0.0      0.00       1.0  
AcceptedCmp4             0.0      0.00       1.0  
AcceptedCmp5             0.0      0.00       1.0  
AcceptedCmp1             0.0      0.00       1.0  
AcceptedCmp2             0.0      0.00       1.0  
Complain                 0.0      0.00       1.0  
Response                 0.0      0.00       1.0

# 4 - Exploration Data Analysis

## 4.1 - `Response` (Kolom target)

## 4.2 - Kolom Numerik

## 4.3 - Kolom Kategorik

## 4.4 - Korelasi Terhadap Kolom `Response`

### 4.4.1 - Kolom Numerik

Menggunakan Kendalls sebab input data merupakan data numerik dan output data adalah data kategorik.


In [104]:
num_col = ['Year_Birth','Income','Recency','MntWines','MntFruits','MntMeatProducts','MntFishProducts','MntSweetProducts','MntGoldProds','NumDealsPurchases','NumWebPurchases','NumCatalogPurchases','NumStorePurchases','NumWebVisitsMonth']
for col in num_col:
    corr_tau, pval_k = stats.kendalltau(df[col], df['Response'])
    print("Korelasi Response dengan",col)
    print(f"tau-correlation: {corr_tau:.2f}, p-value: {pval_k}\n")

Korelasi Response dengan Year_Birth
tau-correlation: 0.02, p-value: 0.24524823598394152

Korelasi Response dengan Income
tau-correlation: 0.13, p-value: 4.545470644980928e-13

Korelasi Response dengan Recency
tau-correlation: -0.17, p-value: 3.219745791134469e-20

Korelasi Response dengan MntWines
tau-correlation: 0.16, p-value: 3.536950153920473e-18

Korelasi Response dengan MntFruits
tau-correlation: 0.12, p-value: 4.0924012467355975e-11

Korelasi Response dengan MntMeatProducts
tau-correlation: 0.18, p-value: 2.6813914010287054e-23

Korelasi Response dengan MntFishProducts
tau-correlation: 0.09, p-value: 4.649823839937091e-07

Korelasi Response dengan MntSweetProducts
tau-correlation: 0.11, p-value: 6.809425480011974e-10

Korelasi Response dengan MntGoldProds
tau-correlation: 0.16, p-value: 2.87194569105795e-18

Korelasi Response dengan NumDealsPurchases
tau-correlation: -0.04, p-value: 0.07188624664034687

Korelasi Response dengan NumWebPurchases
tau-correlation: 0.15, p-value: 1.9

### 4.4.2 - Kolom Kategorik

Menggunakan Chi-squared sebab input data merupakan tipe data kategorik dan output data adalah tipe data kategorik.

Dasar pengambilan keputusan:

H0: Tidak ada relasi antara kolom {i} dan kolom `Response`.

H1: Ada relasi antara kolom {i} dan kolom `Response`.

In [105]:
cat_col = ['Education','Marital_Status','Kidhome','Teenhome','AcceptedCmp1','AcceptedCmp2','AcceptedCmp3','AcceptedCmp4','AcceptedCmp5','Complain']

for col in cat_col:
    contingency_table = pd.crosstab(df[col],df['Response'])
    res = stats.chi2_contingency(contingency_table)
    print("Korelasi Response dengan",col)
    print("P-value:",res.pvalue)
    print("Critical value = 0.05")
    if res.pvalue < 0.05:
        print('Tolak H0, H1 Diterima')
        print('Ada relasi\n')
    else:
        print('Gagal menolak H0')
        print('Tidak ada relasi\n')

Korelasi Response dengan Education
P-value: 0.0009155530597102856
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan Marital_Status
P-value: 1.2842963935056625e-08
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan Kidhome
P-value: 0.000999561158250891
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan Teenhome
P-value: 1.2765470414392849e-12
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan AcceptedCmp1
P-value: 7.092250494721227e-40
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan AcceptedCmp2
P-value: 1.887802102708071e-15
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan AcceptedCmp3
P-value: 1.117571105655191e-28
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi Response dengan AcceptedCmp4
P-value: 3.229210633524326e-14
Critical value = 0.05
Tolak H0, H1 Diterima
Ada relasi

Korelasi 